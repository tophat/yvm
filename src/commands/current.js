const path = require('path')

const {
    getRcFileVersion,
    getValidVersionString,
    getVersionInUse,
} = require('../util/version')
const log = require('../util/log')
const { versionRootPath } = require('../util/utils')
const { getCurrentPath, getPathDelimiter, yvmPath } = require('../util/path')

const currentCommand = async ({ path: pathToUse, shell } = {}) => {
    const pathDelimiter = getPathDelimiter(shell)
    const versionInUse = await getVersionInUse()
    if (!versionInUse) {
        log('Sorry, yarn is NOT installed.')
        return 1
    }
    const envPath = pathToUse || getCurrentPath(shell)
    if (!envPath) {
        log('Environment PATH not found.')
        return 1
    }
    for (const element of envPath.split(pathDelimiter)) {
        if (element.startsWith(versionRootPath(yvmPath))) {
            const [pathVersion] = element
                .split(path.sep)
                .map(getValidVersionString)
                .filter(a => a)
            log.info(`matched yvm version: v${pathVersion} in PATH ${element}`)

            const rcVersion = getRcFileVersion()
            log(`Currently on yarn version ${pathVersion}`)
            if (rcVersion !== null) {
                if (pathVersion === rcVersion) {
                    log(
                        'Your .yvmrc version matches your PATH version, good job!',
                    )
                } else {
                    log(
                        `Your .yvmrc version: ${rcVersion} and PATH version: ${pathVersion} don't match :(`,
                    )
                }
            }
            return 0
        }
    }
    log('Yarn was NOT installed by yvm')
    return 2
}

module.exports = currentCommand
