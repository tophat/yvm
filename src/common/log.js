const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

function log(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

module.exports = log
