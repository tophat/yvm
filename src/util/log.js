const chalk = require('chalk')

const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

const log = (...args) => log.default(...args)

log.default = function defaultLog(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

log.info = function infoLog(...args) {
    if (process.env.YVM_VERBOSE) {
        log(...args)
    }
}

log.error = function errorLog(...args) {
    log(chalk.red(...args))
}

log.success = function successLog(...args) {
    log(chalk.green(...args))
}

log.notice = function noticeLog(...args) {
    log(chalk.yellow(...args))
}

module.exports = log
