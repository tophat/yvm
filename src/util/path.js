const os = require('os')
const path = require('path')

const yvmPath = process.env.YVM_DIR || path.resolve(os.homedir(), '.yvm')
module.exports = { yvmPath }
