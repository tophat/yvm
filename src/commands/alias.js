import log from '../util/log'
import { getFormatter, resolveMatchingAliases, setAlias } from '../util/alias'
import {
    getVersionInUse,
    getYarnVersions,
    resolveVersion,
} from '../util/version'
import { getVersionsFromTags } from '../util/utils'

export const alias = async (nameOrPattern, maybeVersion) => {
    try {
        const format = getFormatter(
            await getVersionsFromTags(),
            getYarnVersions(),
            await getVersionInUse(),
        )

        const safeResolveVersion = async versionString =>
            resolveVersion({ versionString }).catch(e => log.info(e.message))

        if (typeof maybeVersion === 'string') {
            const name = nameOrPattern
            const version = maybeVersion
            let targetVersion
            if (await setAlias({ name, version })) {
                targetVersion = await safeResolveVersion(version)
            }
            const message = format(name, version, targetVersion)
            if (targetVersion) {
                log(message)
                return 0
            } else {
                log.error(message)
                return 1
            }
        } else {
            const pattern = nameOrPattern || ''
            const matchingAliases = await resolveMatchingAliases(pattern)
            for (const {
                name,
                value: { value, version: versionString },
            } of matchingAliases) {
                const version = await safeResolveVersion(versionString)
                log(format(name, value, version))
            }
            const noMatchFound =
                pattern.length > 0 && matchingAliases.length < 1
            return noMatchFound ? 1 : 0
        }
    } catch (e) {
        log.error(e.message)
        return 2
    }
}
