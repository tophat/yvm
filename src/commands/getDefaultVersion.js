import log from 'util/log'
import { getDefaultVersion as getDefault } from 'util/version'

export const getDefaultVersion = async () => {
    try {
        const version = await getDefault()
        if (version) {
            log.capturable(version)
            return 0
        }
        log('No default yarn version set')
        return 1
    } catch (e) {
        log.error(e.message)
        log.info(e.stack)
        return 2
    }
}
