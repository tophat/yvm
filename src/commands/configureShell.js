import fs from 'fs-extra'
import os from 'os'
import path from 'path'

import escapeRegExp from 'lodash/escaperegexp'

import log from '../util/log'
import bashScript from '!!raw-loader!shell/yvm.sh'
import fishScript from '!!raw-loader!shell/yvm.fish'
import yarnShim from '!!raw-loader!shell/yarn_shim.js'

/**
 * Helper utility for unpacking an executable script
 * from yvm.js into a target file. Is a no-op if the
 * target file already exists.
 */
const unpackShellScript = (content, filename) => {
    if (fs.existsSync(filename)) return false
    fs.writeFileSync(filename, content, {
        encoding: 'utf8',
        mode: 0o755,
    })
}

export async function ensureConfig(configFile, configLines) {
    if (!fs.existsSync(configFile)) return false
    let contents = fs.readFileSync(configFile, 'utf8')
    const linesAppended = configLines.map(string => {
        const finalString = `\n${string}`
        if (contents.includes(string)) {
            const matchString = new RegExp(`\n.*${escapeRegExp(string)}.*`)
            contents = contents.replace(matchString, finalString)
            return false
        }
        contents += finalString
        return true
    })
    if (linesAppended.some(a => a)) {
        contents += '\n'
    }
    fs.writeFileSync(configFile, contents)
    log.info(`Configured '${configFile}'`)
    return true
}

const yvmDirVarName = 'YVM_DIR'

export const configureBash = async ({ home, yvmDir }) => {
    const bashRcFile = path.join(home, '.bashrc')
    const bashProFile = path.join(home, '.bash_profile')
    const shPath = path.join(yvmDirVarName, 'yvm.sh')
    const bashConfig = [
        `export ${yvmDirVarName}=${yvmDir}`,
        `[ -r $${shPath} ] && . $${shPath}`,
    ]
    unpackShellScript(bashScript, shPath)
    let configured = false
    if (fs.existsSync(bashRcFile)) {
        configured = await ensureConfig(bashRcFile, bashConfig)
    } else if (fs.existsSync(bashProFile)) {
        configured = await ensureConfig(bashProFile, bashConfig)
    }
    if (!configured) {
        log('Unable to configure BASH at', bashRcFile, 'or', bashProFile)
    }
    return configured
}

export const configureFish = async ({ home, yvmDir }) => {
    const configFile = path.join(home, '.config', 'fish', 'config.fish')
    const fishPath = path.join(yvmDirVarName, 'yvm.fish')
    const configured = await ensureConfig(configFile, [
        `set -x ${yvmDirVarName} ${yvmDir}`,
        `. $${fishPath}`,
    ])
    unpackShellScript(fishScript, fishPath)
    if (!configured) {
        log('Unable to configure FISH at', configFile)
    }
    return configured
}

export const configureZsh = async ({ home, yvmDir }) => {
    const shPath = path.join(yvmDirVarName, 'yvm.sh')
    const configFile = path.join(home, '.zshrc')
    const configured = await ensureConfig(configFile, [
        `export ${yvmDirVarName}=${yvmDir}`,
        `[ -r $${shPath} ] && . $${shPath}`,
    ])
    unpackShellScript(bashScript, shPath)
    if (!configured) {
        log('Unable to configure ZSH at', configFile)
    }
    return configured
}

export const configureShim = async ({ yvmDir }) => {
    const shimFile = path.join(yvmDir, 'shim', 'yarn')
    unpackShellScript(yarnShim, shimFile)
}

export const configureShell = async ({ home, shim, shell = '' } = {}) => {
    try {
        const userHome = home || process.env.HOME || os.homedir()
        const yvmDir =
            process.env.YVM_INSTALL_DIR || path.join(userHome, '.yvm')
        const configHandlers = {
            bash: configureBash,
            fish: configureFish,
            zsh: configureZsh,
        }
        const updatingShellConfigs = []
        if (shim) updatingShellConfigs.push(configureShim({ yvmDir }))
        for (const supportedShell of Object.keys(configHandlers)) {
            if (supportedShell.includes(shell)) {
                updatingShellConfigs.push(
                    configHandlers[supportedShell]({ home: userHome, yvmDir }),
                )
            }
        }
        const result = await Promise.all(updatingShellConfigs)
        const allSuccessful = result.every(a => a)
        const anySuccessful = result.some(a => a)
        const isSuccessful = allSuccessful || (anySuccessful && !shell)
        if (!isSuccessful) {
            return 1
        }
        log('Shell configured successfully')
        return 0
    } catch (e) {
        log(e.message)
        log.info(e.stack)
        return 2
    }
}
