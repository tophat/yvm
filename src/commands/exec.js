import path from 'path'
import { execFileSync } from 'child_process'

import { ensureVersionInstalled } from './install'
import { getSplitVersionAndArgs } from '../util/version'
import log from '../util/log'
import { yvmPath } from '../util/path'

const getYarnPath = (version, rootPath) =>
    path.resolve(rootPath, `versions/v${version}`)

/**
 * __WARNING__
 * When changing this logic ensure that passing of stdio works correctly.
 * e.g. user input, arrows keys, etc.
 *
 * __TEST__
 * - `nvm use 8.0.0` (lowest supported node version)
 * - `make install`
 * - `yvm exec contributors:add` and go through the steps
 */
const runYarn = (version, extraArgs, rootPath = yvmPath) => {
    process.argv = ['', ''].concat(extraArgs)
    const filePath = path.resolve(getYarnPath(version, rootPath), 'bin/yarn.js')
    const command = `${filePath} ${extraArgs.join(' ')}`
    log.info(command)
    try {
        execFileSync(filePath, extraArgs, {
            stdio: 'inherit',
        })
    } catch (error) {
        log('yarn failed, non-zero exit')
        throw error
    }
}

export const exec = async (maybeVersion, rest = []) => {
    try {
        const [version, args] = await getSplitVersionAndArgs(
            maybeVersion,
            ...rest,
        )
        await ensureVersionInstalled(version)
        await runYarn(version, args)
        return 0
    } catch (e) {
        log(e.message)
        log.info(e.stack)
        return 1
    }
}
