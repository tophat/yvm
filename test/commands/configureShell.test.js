import fs from 'fs-extra'
import path from 'path'
import os from 'os'
import mockProps from 'jest-mock-props'
mockProps.extend(jest)

import log from 'util/log'
import { configureShell, ensureConfig } from 'commands/configureShell'

describe('configureShell', () => {
    const mockHomeValue = 'config-mock-home'
    const envHomeMock = jest.spyOnProp(process.env, 'HOME')
    const mockInstallDir = 'config-mock-install-dir/.yvm'
    const envYvmInstallDir = jest.spyOnProp(process.env, 'YVM_INSTALL_DIR')
    jest.spyOn(os, 'homedir')
    jest.spyOn(log, 'default')
    jest.spyOn(log, 'info')

    const fileToPath = ([file]) => path.join(mockHomeValue, file)
    const rcFiles = {
        bashrc: fileToPath`.bashrc`,
        bashpro: fileToPath`.bash_profile`,
        zshrc: fileToPath`.zshrc`,
        fishconf: fileToPath`.config/fish/config.fish`,
    }
    const rcFilePaths = Object.values(rcFiles)
    const confirmShellConfig = () => {
        const configs = rcFilePaths.reduce((configs, file) => {
            const content = fs.existsSync(file) && fs.readFileSync(file, 'utf8')
            return Object.assign(configs, { [file]: content || null })
        }, {})
        expect(configs).toMatchSnapshot()
    }

    beforeEach(() => {
        jest.clearAllMocks()
        fs.mkdirSync(mockHomeValue)
        rcFilePaths.forEach(filePath => fs.outputFileSync(filePath, 'dummy'))
    })

    afterEach(() => {
        envHomeMock.mockReset()
        envYvmInstallDir.mockReset()
        fs.removeSync(mockHomeValue)
        fs.removeSync('config-mock-install-dir')
    })

    afterAll(jest.restoreAllMocks)

    it('configures only bashrc', async () => {
        envYvmInstallDir.mockValue(mockInstallDir)
        expect(
            await configureShell({ home: mockHomeValue, shell: 'bash' }),
        ).toBe(0)
        confirmShellConfig()
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining('Configured'),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.bashrc),
        )
        expect(log.default).toHaveBeenCalledWith(
            'Shell configured successfully',
        )
    })

    it('configures only bash_profile when no bashrc', async () => {
        envYvmInstallDir.mockValue(mockInstallDir)
        fs.removeSync(rcFiles.bashrc)
        expect(
            await configureShell({ home: mockHomeValue, shell: 'bash' }),
        ).toBe(0)
        confirmShellConfig()
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining('Configured'),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.bashpro),
        )
        expect(log.default).toHaveBeenCalledWith(
            'Shell configured successfully',
        )
    })

    it('configures only fish', async () => {
        envHomeMock.mockValue(mockHomeValue)
        expect(await configureShell({ shell: 'fish' })).toBe(0)
        confirmShellConfig()
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining('Configured'),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.fishconf),
        )
        expect(log.default).toHaveBeenCalledWith(
            'Shell configured successfully',
        )
    })

    it('configures only zsh', async () => {
        envHomeMock.mockValue()
        os.homedir.mockReturnValueOnce(mockHomeValue)
        expect(await configureShell({ shell: 'zsh' })).toBe(0)
        confirmShellConfig()
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining('Configured'),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.zshrc),
        )
        expect(log.default).toHaveBeenCalledWith(
            'Shell configured successfully',
        )
    })

    it('configures all', async () => {
        expect(await configureShell({ home: mockHomeValue })).toBe(0)
        confirmShellConfig()
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining('Configured'),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.bashrc),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.fishconf),
        )
        expect(log.info).toHaveBeenCalledWith(
            expect.stringContaining(rcFiles.zshrc),
        )
        expect(log.default).toHaveBeenCalledWith(
            'Shell configured successfully',
        )
    })

    it('configures none', async () => {
        fs.emptyDirSync(mockHomeValue)
        expect(await configureShell({ home: mockHomeValue })).toBe(1)
        rcFilePaths.forEach(file => expect(fs.existsSync(file)).toBe(false))
        confirmShellConfig()
        expect(log.default.mock.calls).toMatchSnapshot()
    })

    it('handle error', async () => {
        const mockError = new Error('mock error')
        os.homedir.mockImplementationOnce(() => {
            throw mockError
        })
        envHomeMock.mockValue()
        expect(await configureShell()).toBe(2)
        expect(log.default).toHaveBeenCalledWith(mockError.message)
        expect(log.info).toHaveBeenCalledWith(mockError.stack)
    })

    describe('ensureConfig', () => {
        it('updates exiting lines', async () => {
            const fileName = 'some-random-rc'
            const initial = `with
# -- specific existing line --
that should be replaced`
            const expected = `with
specific existing line
that should be replaced`
            fs.outputFileSync(fileName, initial)
            await ensureConfig(fileName, ['specific existing line'])
            expect(fs.readFileSync(fileName, 'utf8')).toEqual(expected)
            fs.unlinkSync(fileName)
        })
    })
})
