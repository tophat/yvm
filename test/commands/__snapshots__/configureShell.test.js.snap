// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configureShell configures all 1`] = `
Object {
  "config-mock-home/.bash_profile": "dummy",
  "config-mock-home/.bashrc": "dummy
export YVM_DIR=config-mock-home/.yvm
[ -r $YVM_DIR/yvm.sh ] && . $YVM_DIR/yvm.sh
",
  "config-mock-home/.config/fish/config.fish": "dummy
set -x YVM_DIR config-mock-home/.yvm
. $YVM_DIR/yvm.fish
",
  "config-mock-home/.zshrc": "dummy
export YVM_DIR=config-mock-home/.yvm
[ -r $YVM_DIR/yvm.sh ] && . $YVM_DIR/yvm.sh
",
}
`;

exports[`configureShell configures none 1`] = `
Object {
  "config-mock-home/.bash_profile": null,
  "config-mock-home/.bashrc": null,
  "config-mock-home/.config/fish/config.fish": null,
  "config-mock-home/.zshrc": null,
}
`;

exports[`configureShell configures none 2`] = `
Array [
  Array [
    "Unable to configure BASH at",
    "config-mock-home/.bashrc",
    "or",
    "config-mock-home/.bash_profile",
  ],
  Array [
    "Unable to configure FISH at",
    "config-mock-home/.config/fish/config.fish",
  ],
  Array [
    "Unable to configure ZSH at",
    "config-mock-home/.zshrc",
  ],
]
`;

exports[`configureShell configures only bash_profile when no bashrc 1`] = `
Object {
  "config-mock-home/.bash_profile": "dummy
export YVM_DIR=config-mock-install-dir/.yvm
[ -r $YVM_DIR/yvm.sh ] && . $YVM_DIR/yvm.sh
",
  "config-mock-home/.bashrc": null,
  "config-mock-home/.config/fish/config.fish": "dummy",
  "config-mock-home/.zshrc": "dummy",
}
`;

exports[`configureShell configures only bashrc 1`] = `
Object {
  "config-mock-home/.bash_profile": "dummy",
  "config-mock-home/.bashrc": "dummy
export YVM_DIR=config-mock-install-dir/.yvm
[ -r $YVM_DIR/yvm.sh ] && . $YVM_DIR/yvm.sh
",
  "config-mock-home/.config/fish/config.fish": "dummy",
  "config-mock-home/.zshrc": "dummy",
}
`;

exports[`configureShell configures only fish 1`] = `
Object {
  "config-mock-home/.bash_profile": "dummy",
  "config-mock-home/.bashrc": "dummy",
  "config-mock-home/.config/fish/config.fish": "dummy
set -x YVM_DIR config-mock-home/.yvm
. $YVM_DIR/yvm.fish
",
  "config-mock-home/.zshrc": "dummy",
}
`;

exports[`configureShell configures only zsh 1`] = `
Object {
  "config-mock-home/.bash_profile": "dummy",
  "config-mock-home/.bashrc": "dummy",
  "config-mock-home/.config/fish/config.fish": "dummy",
  "config-mock-home/.zshrc": "dummy
export YVM_DIR=config-mock-home/.yvm
[ -r $YVM_DIR/yvm.sh ] && . $YVM_DIR/yvm.sh
",
}
`;
